backends/transforms/utils.py

```diff
@@ -187,3 +230,128 @@ def delete_constant_placeholder(exp_program: ExportedProgram, node: torch.fx.Nod
+    else:
+        target = name
+
+    # Check if target already exists
+    if target in exp_program.state_dict:
+        raise RuntimeError(f"Buffer target '{target}' already exists in state_dict")
+
+    persistent_buffer = True
+    exp_program.state_dict[target] = data
+
+    # Create fake tensor using helper function
+    fake_tensor = _get_fake_tensor_mode(graph, data)
+
+    # Signature ordering is as follows:
+    # Inputs = [*parameters_buffers_constant_tensors, *flattened_user_inputs]
+    #                       ^^^^^^^
+    # Outputs = [*mutated_inputs, *flattened_user_outputs]
+
+    # Validate that the signature is up to date and aligned with the placeholders in the graph
+    placeholders = [n for n in graph.nodes if n.op == "placeholder"]
+    if len(placeholders) != len(exp_program.graph_signature.input_specs):
+        raise RuntimeError(
+            f"Graph has {len(placeholders)} placeholder nodes but signature has "
+            f"{len(exp_program.graph_signature.input_specs)} input specs"
+        )
+
+    for node, input_spec in zip(placeholders, exp_program.graph_signature.input_specs):
+        if node.name != input_spec.arg.name:
+            raise RuntimeError(
+                f"Input node {node.name} does not match input spec {input_spec.arg.name}"
+            )
+
+    # Find indices of user inputs and constant tensors
+    user_input_indices_or_consts = [
+        i
+        for i, spec in enumerate(exp_program.graph_signature.input_specs)
+        if spec.kind == InputKind.USER_INPUT or spec.kind == InputKind.CONSTANT_TENSOR
+    ]
+
+    if not user_input_indices_or_consts:
+        # If no user inputs or constants exist, insert at the end of all existing inputs
+        node_index = len(exp_program.graph_signature.input_specs)
+        # Create node at the end of existing placeholders
+        node = graph.create_node(op="placeholder", name=name, target=name)
+        node.meta["val"] = fake_tensor
+    else:
+        node_index = user_input_indices_or_consts[0]
+        # Find the first user input placeholder node
+        first_user_input_or_constant = _spec_to_node(
+            exp_program, exp_program.graph_signature.input_specs[node_index]
+        )
+        # Insert the mutable buffer node before the first user input node or constant node
+        with graph.inserting_before(first_user_input_or_constant):
+            # Create node
+            node = graph.create_node(op="placeholder", name=name, target=name)
+            node.meta["val"] = fake_tensor
+
+    # Create input spec for the mutable buffer
+    input_specs = exp_program.graph_signature.input_specs
+    mutable_input_spec = InputSpec(
+        InputKind.BUFFER, TensorArgument(name), target, persistent_buffer
+    )
+    input_specs.insert(node_index, mutable_input_spec)
+
+    # Create output spec for the mutable buffer
+    # We don't have to explicitly update the output.args[0] since it will be updated by the user based on the graph.
```

Comment:
I dont follow this. If you have to return input placeholder node, that you created, because it is a mutation, doesnt the output node's args have to be updated?

Reply:
iiuc your end idea @digantdesai  is to add the buffer and then consume and mutate it with another pass that adds a custom op, so you could just do the work of setting up the outputs when you actually mutate the buffer. 

I agree with Kimish though that I think its a lot easier to compose this pass if it leaves the graph in a valid state rather then implicitly assuming the following pass will fix it up.

Reply:
>IIUC, you are saying after each line in the pass the graph should be in a valid state

I think public utils should probably leave it in a valid state. The graph can obviously have invalid state in the middle of a pass, but if I was joe schmo who didnt really understand the nuances of what this util did and just had the high level goal of "add a buffer to it" I would expect the graph to be sound after that operation.